src/Microsoft.Health.Fhir.Api/Features/Throttling/ThrottlingMiddleware.cs

```diff
@@ -127,35 +137,137 @@ public async Task Invoke(HttpContext context)
                 return;
             }
 
-            try
+            bool canRun = false;
+            bool queueSizeExceeded = false;
+            LinkedListNode<TaskCompletionSource<object>> queueNode = null;
+
+            for (int i = 0; i < 2; i++)
             {
-                if (Interlocked.Increment(ref _requestsInFlight) <= _configuration.ConcurrentRequestLimit)
+                // we do two loop iterations only when we need to queue up the request.
+                lock (_queue)
                 {
-                    // Still within the concurrent request limit, let the request through.
-                    Interlocked.Increment(ref _currentPeriodSuccessCount);
-                    await _next(context);
+                    if (_requestsInFlight < _concurrentRequestLimit)
+                    {
+                        canRun = true;
+                        _requestsInFlight++;
```

Comment:
Does this variable also need to be changed with Interlocked or volatile?

Reply:
We only access in the the scope of the lock, so no need.